Add `tar_terra_vrt()`

brownag opened the pull request:
This PR implements a geotargets wrapper for `terra::vrt()`. The target data stored is an XML file referencing bands and sources that comprise a larger image. The `tar_format()` write method writes an XML file to the target store, the read method creates a SpatRaster object from that XML. 

This ties in with existing `tar_terra_tiles()` function by providing an option for working with tiles processed by dynamic branching that does not require saving the entire set as a single output file / target. This would complement the example in the tar_terra_tiles vignette where `merge()` is called on a SpatRasterCollection target. In contrast to a `tar_terra_sprc()` target, which is storing a copy of the data with each tile as a subdataset, `tar_terra_vrt()` allows re-use of existing tile targets but can be used in essentially the same way as the the merged SpatRaster.

For example:
``` r
library(targets)
library(terra)
#> terra 1.8.9

tar_script({
    library(targets)
    library(geotargets)
    library(terra)
    
    list(
        
        tar_terra_rast(x, rast(system.file("ex", "elev.tif", package = "terra"))), 
        
        tar_terra_tiles(
            xt,
            x,
            \(x) tile_grid(x, ncol = 2, nrow = 2)
        ),
        
        tar_terra_sprc(
          s,
          sprc(xt)
        ),
        
        tar_terra_vrt(
            v.vrt,
            xt
        ),
        tar_target(
            vrt_xml,
            readLines(tar_path_target(v.vrt))
        )
    )
    
})

tar_make()
#> terra 1.8.9
#> ▶ dispatched target x
#> ● completed target x [0.009 seconds, 8.523 kilobytes]
#> ▶ dispatched target xt_exts
#> ● completed target xt_exts [0.008 seconds, 146 bytes]
#> ▶ dispatched branch xt_9949683a4fca725a
#> ● completed branch xt_9949683a4fca725a [0.023 seconds, 3.629 kilobytes]
#> ▶ dispatched branch xt_5c0e4ea6d54a188b
#> ● completed branch xt_5c0e4ea6d54a188b [0.064 seconds, 1.933 kilobytes]
#> ▶ dispatched branch xt_97f5221e9008d688
#> ● completed branch xt_97f5221e9008d688 [0.007 seconds, 3.397 kilobytes]
#> ▶ dispatched branch xt_ba4129c8b49fb340
#> ● completed branch xt_ba4129c8b49fb340 [0.007 seconds, 3.425 kilobytes]
#> ● completed pattern xt 
#> ▶ dispatched target v.vrt
#> ● completed target v.vrt [0 seconds, 2.426 kilobytes]
#> ▶ dispatched target s
#> ● completed target s [0.025 seconds, 12.363 kilobytes]
#> ▶ dispatched target vrt_xml
#> ● completed target vrt_xml [0.001 seconds, 743 bytes]
#> ▶ ended pipeline [0.603 seconds]

tar_load(x)
x
#> class       : SpatRaster 
#> dimensions  : 90, 95, 1  (nrow, ncol, nlyr)
#> resolution  : 0.008333333, 0.008333333  (x, y)
#> extent      : 5.741667, 6.533333, 49.44167, 50.19167  (xmin, xmax, ymin, ymax)
#> coord. ref. : lon/lat WGS 84 (EPSG:4326) 
#> source      : x 
#> name        : elevation 
#> min value   :       141 
#> max value   :       547

tar_load(v.vrt)
plot(v.vrt)
```

![](https://i.imgur.com/NaZyvz3.png)<!-- -->

``` r
v.vrt
#> class       : SpatRaster 
#> dimensions  : 90, 95, 1  (nrow, ncol, nlyr)
#> resolution  : 0.008333333, 0.008333333  (x, y)
#> extent      : 5.741667, 6.533333, 49.44167, 50.19167  (xmin, xmax, ymin, ymax)
#> coord. ref. : lon/lat WGS 84 (EPSG:4326) 
#> source      : v.vrt 
#> name        :   v 
#> min value   : 141 
#> max value   : 547

tar_load(vrt_xml)
cat(vrt_xml, sep = "\n")
#> <VRTDataset rasterXSize="95" rasterYSize="90">
#>   <SRS dataAxisToSRSAxisMapping="2,1">GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AXIS["Latitude",NORTH],AXIS["Longitude",EAST],AUTHORITY["EPSG","4326"]]</SRS>
#>   <GeoTransform>  5.7416666666666663e+00,  8.3333333333333350e-03,  0.0000000000000000e+00,  5.0191666666666663e+01,  0.0000000000000000e+00, -8.3333333333333332e-03</GeoTransform>
#>   <VRTRasterBand dataType="Int16" band="1">
#>     <NoDataValue>-32768</NoDataValue>
#>     <ColorInterp>Gray</ColorInterp>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/xt_9949683a4fca725a</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="48" RasterYSize="45" DataType="Int16" BlockXSize="48" BlockYSize="45" />
#>       <SrcRect xOff="0" yOff="0" xSize="48" ySize="45" />
#>       <DstRect xOff="0" yOff="0" xSize="48" ySize="45" />
#>       <NODATA>-32768</NODATA>
#>     </ComplexSource>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/xt_5c0e4ea6d54a188b</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="47" RasterYSize="45" DataType="Int16" BlockXSize="47" BlockYSize="45" />
#>       <SrcRect xOff="0" yOff="0" xSize="47" ySize="45" />
#>       <DstRect xOff="48" yOff="0" xSize="47" ySize="45" />
#>       <NODATA>-32768</NODATA>
#>     </ComplexSource>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/xt_97f5221e9008d688</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="48" RasterYSize="45" DataType="Int16" BlockXSize="48" BlockYSize="45" />
#>       <SrcRect xOff="0" yOff="0" xSize="48" ySize="45" />
#>       <DstRect xOff="0" yOff="45" xSize="48" ySize="45" />
#>       <NODATA>-32768</NODATA>
#>     </ComplexSource>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/xt_ba4129c8b49fb340</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="47" RasterYSize="45" DataType="Int16" BlockXSize="47" BlockYSize="45" />
#>       <SrcRect xOff="0" yOff="0" xSize="47" ySize="45" />
#>       <DstRect xOff="48" yOff="45" xSize="47" ySize="45" />
#>       <NODATA>-32768</NODATA>
#>     </ComplexSource>
#>   </VRTRasterBand>
#> </VRTDataset>
```

Aariq commented:
I'm eager to test this out once I have a decent chunk of time!  I think I'll end up using this in some workflows.

brownag commented:
> I'm eager to test this out once I have a decent chunk of time! I think I'll end up using this in some workflows.

Cool. One thing I am realizing now is that the marshal/unmarshal methods likely need some attention considering the (currently allowed) heterogeneous input object types. I didn't think that through fully when I wrote this up--so I am open to any suggestions on improving that in particular.

brownag commented:
Following https://github.com/ropensci/geotargets/pull/138#issuecomment-2605972811 I've updated the behavior of `tar_terra_vrt()` so that one can no longer use a list of SpatRaster or a character vector of paths as input. I changed this because the marshal/unmarshal methods (namely `terra::wrap()`) will not work on list or character vector. 

Now to use the output from e.g. `tar_terra_tiles()` (as shown in the tests) we need to wrap the tile result in a call to `terra::sprc()` to create a SpatRasterCollection and then pass that SPRC object to `tar_terra_vrt()`. To use a character vector of paths, first call `terra::vrt()` on the path vector to return a SpatRaster and then pass that object (as shown in the example)

I also added support for SpatRasterDataset as input. 

Essentially, terra methods `rast()` or `vrt()`, `sds()`, and `sprc()` all produce "wrappable" objects that can be marshalled. As they all have a `terra::sources()` method defined, they can be supported equivalently in `tar_terra_vrt()`. Since the "read" method pulls from the VRT file, it is fine to use just `rast()` there, even though the input object from `command` could have been any one of the 3 SpatRaster* types.

Aariq commented:
This works like a charm for bringing tiles back together.  This is maybe more of a VRT question than a question about this target constructor, but what does `vrt(sds(raster_list))` even do?  E.g. this example gives me a raster with a single layer for r3, which is not what I expected.  However, I can replicate this with `vrt(sources(sds(r2)))`, so I guess I just don't understand when you'd pass an `sds()` to `vrt()` and if the answer is "almost never", should we just make the `sprc()` wrapper "built-in"?

```r
targets::tar_script({
    library(targets)
    library(geotargets)
    list(
        tar_terra_rast(r, terra::rast(
            system.file("ex", "elev.tif", package = "terra")
        )),
        tar_target(mult, 1:5),
        tar_terra_rast(r2, r * mult, pattern = map(mult)),
        tar_terra_vrt(r3, terra::sds(r2))
    )
})
```

brownag commented:
@Aariq Great point. I think to get the behavior you are expecting with a SpatRasterDataset where all the source subdatasets are in separate files but overlap, you need to pass the option `-separate` to gdalbuildvrt via `vrt(options=)` to make each source a separate VRTRasterBand. We can do this in `tar_terra_vrt()` currently through `...` passed through to `vrt()` (may want to make VRT options an explicit argument). The default behavior writes _all four sources_ into _one_ band, which looks like just the first source provided all subdatsets have the same grid system. See example below.

A more "realistic" use of the SpatRasterDataset might have `mult` be some multiplicative factor on e.g. raster resolution for the same extent. Then the extents would match and overlap, but the grid would be different, and the frankenband created without `-separate` would look different from the first band.

To your point about "almost never": I think that SpatRasterDataset is rarely used in general--it is helpful for bookkeeping but does not provide many advantages over a list of SpatRaster. I included it simply because it has a `sources()` method.

I would be fine giving preference to the SpatRasterCollection case and including some internal handling to ease use with `tar_terra_tiles()`.  I think if we wanted to provide generic support for SpatRasterDataset we would need to also account for heterogeneous grids in the source bands... because while `rast()` will read them into a single SpatRaster dealing with bands that have different resolution is not properly handled. While it is technically legitimate to have overlapping "subdatasets" as separate bands in a VRT, it would need to be read back in band by band with `rast()`+`sds()`. An instance where one might want to do this is in calculating some sort of derived pixel function within the VRT: https://gdal.org/en/stable/drivers/raster/vrt.html#using-derived-bands-with-pixel-functions-in-c-c

``` r
library(targets)
library(terra)
#> terra 1.8.29

tar_script({
    library(targets)
    library(geotargets)
    list(
        tar_terra_rast(r, terra::rast(
            system.file("ex", "elev.tif", package = "terra")
        )),
        tar_target(mult, 1:5),
        tar_terra_rast(r2, r * mult, pattern = map(mult)),
        tar_terra_vrt(r3_1, terra::sds(r2)),
        tar_terra_vrt(r3_2, terra::sds(r2), options = "-separate")
    )
})

tar_make()
#> ▶ dispatched target mult
#> ● completed target mult [0.001 seconds, 99 bytes]
#> ▶ dispatched target r
#> ● completed target r [0.005 seconds, 8.523 kilobytes]
#> ▶ dispatched branch r2_6c76ae3544ef89d8
#> ● completed branch r2_6c76ae3544ef89d8 [0.004 seconds, 10.471 kilobytes]
#> ▶ dispatched branch r2_3e082e9c41fcf968
#> ● completed branch r2_3e082e9c41fcf968 [0.003 seconds, 10.237 kilobytes]
#> ▶ dispatched branch r2_3913cc3fc7230836
#> ● completed branch r2_3913cc3fc7230836 [0.003 seconds, 10.709 kilobytes]
#> ▶ dispatched branch r2_006b9913d4c45d0e
#> ● completed branch r2_006b9913d4c45d0e [0.002 seconds, 9.865 kilobytes]
#> ▶ dispatched branch r2_f83391952de6e967
#> ● completed branch r2_f83391952de6e967 [0.002 seconds, 10.847 kilobytes]
#> ● completed pattern r2 
#> ▶ dispatched target r3_1
#> ● completed target r3_1 [0.001 seconds, 2.876 kilobytes]
#> ▶ dispatched target r3_2
#> ● completed target r3_2 [0.01 seconds, 3.264 kilobytes]
#> ▶ ended pipeline [0.254 seconds]

v1 <- tar_path_target(r3_1)
v2 <- tar_path_target(r3_2)

rv1 <- tar_read(r3_1)
rv2 <- tar_read(r3_2)

plot(rv1)
```

![](https://i.imgur.com/UCZ5rcx.png)<!-- -->

``` r
cat(readLines(v1), sep = "\n")
#> <VRTDataset rasterXSize="95" rasterYSize="90">
#>   <SRS dataAxisToSRSAxisMapping="2,1">GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AXIS["Latitude",NORTH],AXIS["Longitude",EAST],AUTHORITY["EPSG","4326"]]</SRS>
#>   <GeoTransform>  5.7416666666666663e+00,  8.3333333333333367e-03,  0.0000000000000000e+00,  5.0191666666666663e+01,  0.0000000000000000e+00, -8.3333333333333332e-03</GeoTransform>
#>   <VRTRasterBand dataType="UInt32" band="1">
#>     <NoDataValue>4294967295</NoDataValue>
#>     <ColorInterp>Gray</ColorInterp>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_6c76ae3544ef89d8</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_3e082e9c41fcf968</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_3913cc3fc7230836</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_006b9913d4c45d0e</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_f83391952de6e967</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>   </VRTRasterBand>
#> </VRTDataset>

plot(rv2)
```

![](https://i.imgur.com/3tHedhf.png)<!-- -->

``` r
cat(readLines(v2), sep = "\n")
#> <VRTDataset rasterXSize="95" rasterYSize="90">
#>   <SRS dataAxisToSRSAxisMapping="2,1">GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AXIS["Latitude",NORTH],AXIS["Longitude",EAST],AUTHORITY["EPSG","4326"]]</SRS>
#>   <GeoTransform>  5.7416666666666663e+00,  8.3333333333333367e-03,  0.0000000000000000e+00,  5.0191666666666663e+01,  0.0000000000000000e+00, -8.3333333333333332e-03</GeoTransform>
#>   <VRTRasterBand dataType="UInt32" band="1">
#>     <NoDataValue>4294967295</NoDataValue>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_6c76ae3544ef89d8</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>   </VRTRasterBand>
#>   <VRTRasterBand dataType="UInt32" band="2">
#>     <NoDataValue>4294967295</NoDataValue>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_3e082e9c41fcf968</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>   </VRTRasterBand>
#>   <VRTRasterBand dataType="UInt32" band="3">
#>     <NoDataValue>4294967295</NoDataValue>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_3913cc3fc7230836</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>   </VRTRasterBand>
#>   <VRTRasterBand dataType="UInt32" band="4">
#>     <NoDataValue>4294967295</NoDataValue>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_006b9913d4c45d0e</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>   </VRTRasterBand>
#>   <VRTRasterBand dataType="UInt32" band="5">
#>     <NoDataValue>4294967295</NoDataValue>
#>     <ComplexSource>
#>       <SourceFilename relativeToVRT="0">_targets/objects/r2_f83391952de6e967</SourceFilename>
#>       <SourceBand>1</SourceBand>
#>       <SourceProperties RasterXSize="95" RasterYSize="90" DataType="UInt32" BlockXSize="95" BlockYSize="21" />
#>       <SrcRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <DstRect xOff="0" yOff="0" xSize="95" ySize="90" />
#>       <NODATA>4294967295</NODATA>
#>     </ComplexSource>
#>   </VRTRasterBand>
#> </VRTDataset>
```

Aariq commented:
I'm ok with how things work as-is (because wrapping in `sprc()` is well documented) if you think it is appropriate to allow that flexibility!  Just wanted to make sure we weren't allowing/encouraging something that is always a bad idea

Aariq reviewed (changes requested):
I thought of one other thing:  Can you add a @seealso link to `tar_terra_vrt()` in the docs for `tar_terra_tiles()`?

brownag commented:
> I thought of one other thing: Can you add a @Seealso link to `tar_terra_vrt()` in the docs for `tar_terra_tiles()`?

Good idea, done in 6c8065c

njtierney commented:
Looks good to me! Just a small thing - could you add a NEWS bullet for this?

brownag commented:
> Looks good to me! Just a small thing - could you add a NEWS bullet for this?

Great. Sure, done in https://github.com/ropensci/geotargets/pull/138/commits/1711b11292cf4b111c936c479c2f7adb79932620 feel free to reword as needed

Aariq commented:
Thanks!  The news bullet should actually go in the *next* version, but @njtierney can fix that after completing #163
